kernel/kernel-api/src/main/java/io/delta/kernel/internal/commitrange/CommitRangeFactory.java

```diff
@@ -0,0 +1,145 @@
+  private final Path logPath;
+
+  CommitRangeFactory(Engine engine, CommitRangeBuilderImpl.Context ctx) {
+    this.ctx = ctx;
+    this.tablePath = new Path(resolvePath(engine, ctx.unresolvedPath));
+    this.logPath = new Path(tablePath, "_delta_log");
+  }
+
+  CommitRangeImpl create(Engine engine) {
+    long startVersion = resolveStartVersion(engine);
+    Optional<Long> endVersionOpt = resolveEndVersionIfSpecified(engine);
+    validateVersionRange(startVersion, endVersionOpt);
+    logResolvedVersions(startVersion, endVersionOpt);
+    // TODO: for now we just store a list of commit files, this will be updated when we add support
+    //  for ccv2
+    List<FileStatus> deltaFiles =
+        DeltaLogActionUtils.getCommitFilesForVersionRange(
+            engine, tablePath, startVersion, endVersionOpt);
+    // Once we have listed files, we can resolve endVersion=latestVersion for the default case
+    long endVersion = endVersionOpt.orElseGet(() -> extractLatestVersion(deltaFiles));
+    if (!endVersionOpt.isPresent()) {
+      logger.info("{}: Resolved endVersion={} to the latest version", tablePath, endVersion);
+    }
+    return new CommitRangeImpl(
+        tablePath, ctx.startBoundaryOpt, ctx.endBoundaryOpt, startVersion, endVersion, deltaFiles);
+  }
+
+  private long resolveStartVersion(Engine engine) {
+    if (!ctx.startBoundaryOpt.isPresent()) {
+      // Default to version 0 if no start boundary is provided
+      return 0L;
+    }
+    CommitRangeBuilder.CommitBoundary startBoundary = ctx.startBoundaryOpt.get();
+
+    if (startBoundary.isVersion()) {
+      return startBoundary.getVersion();
+    } else {
+      logger.info(
+          "{}: Trying to resolve start-boundary timestamp {} to version",
+          tablePath,
+          startBoundary.getTimestamp());
+      // TODO: support ccv2 tables
+      return DeltaHistoryManager.getVersionAtOrAfterTimestamp(
+          engine,
+          logPath,
+          startBoundary.getTimestamp(),
+          (SnapshotImpl) startBoundary.getLatestSnapshot());
+    }
+  }
+
+  /**
+   * This method resolves the endBoundary to a version if it is specified. For a version-based
+   * boundary, this just returns the version. For a timestamp-based boundary, this resolves the
+   * timestamp to version. When the boundary is not specified, this returns empty, as the version
+   * cannot be resolved until later after we have performed any listing.
+   */
+  private Optional<Long> resolveEndVersionIfSpecified(Engine engine) {
+    if (!ctx.endBoundaryOpt.isPresent()) {
+      // When endBoundary is not provided, we default to the latest version. We cannot resolve the
+      // latest version until later after we have performed any listing.
+      return Optional.empty();
+    }
+    CommitRangeBuilder.CommitBoundary endBoundary = ctx.endBoundaryOpt.get();
+
+    if (endBoundary.isVersion()) {
+      return Optional.of(endBoundary.getVersion());
+    } else {
+      logger.info(
```

Comment:
could this log be pushed down into `getVersionBeforeOrAtTimestamp` ? same for the similar invocation above?

Reply:
I think that potentially has less information than here (i.e. we maybe don't want to hard-code that log to be resolving "end-boundary" specifically.)

It seems like there could be better logging in `DeltaHistoryManager` though -- I can add some better logging there in a separate PR if that helps?

Reply:
SGTM